src/herring/tl_helpers.hpp

```diff
@@ -0,0 +1,305 @@
+#pragma once
+
+#include <exception>
+#include <iostream>
+#include <stack>
+#include <string>
+#include <variant>
+
+#include <treelite/tree.h>
+
+#include <herring/model.hpp>
+#include <herring/output_ops.hpp>
+
+namespace herring {
+
+struct unconvertible_model_exception : std::exception {
+  unconvertible_model_exception () : msg_{"Model could not be converted"}
+  {
+  }
+
+  unconvertible_model_exception (std::string msg) : msg_{msg}
+  {
+  }
+
+  unconvertible_model_exception (char const* msg) : msg_{msg}
+  {
+  }
+
+  virtual char const* what() const noexcept { return msg_.c_str(); }
+ private:
+  std::string msg_;
+};
+
+// TODO(wphicks): Currently, the model use_inclusive_threshold parameter is
+// changed as a side-effect of this function. This is messy and confusing, and
+// it should be fixed in a later refactor
+template<typename tree_t, typename tl_threshold_t, typename tl_output_t>
+auto convert_tree(treelite::Tree<tl_threshold_t, tl_output_t> const& tl_tree, bool& use_inclusive_threshold) {
+  auto result = tree_t{};
+  result.nodes.reserve(tl_tree.num_nodes);
+  result.default_distant.reserve(tl_tree.num_nodes);
+
+  // TL node id for current node
+  auto cur_node_id = int{};
+  // TL node id for hot child of current node
+  auto hot_child = int{};
+  // TL node id for distant child of current node
+  auto distant_child = int{};
+
+  // Stack of TL node ids for DFS
+  auto node_stack = std::stack<int, std::vector<int>>{};
+  // Keep track of final location of parent in nodes vector for each node currently in stack
+  auto parent_stack = std::stack<std::size_t, std::vector<std::size_t>>{};
+  // TODO(wphicks): Just store a reference to the parent directly rather than
```

Comment:
For smaller cleanups, perhaps a tracking gh issue and reference to it in the code comments?